File: algorithms/LinkedList.cpp

node addNode(node head, int value) {
    node temp, p;
    temp = createNode();
    temp->data = value;
    if(head == NULL) {
        head = temp;
    } else {
        p = head;
        while (p->next != NULL) {
            p=p->next;
        }
        p->next = temp;
    }
    return head;
}

bool Contains(node head, int value) {
    node p;
    p = head;
    while(p != null && p->data != value){
        p = p->next;
    }
    if(p==NULL) return false;
    return true;
}

node Prepend(node head, node tail, int value) {
    node temp;
    temp = createNode(value);
    temp.next = head;
    head = temp;
    if(tail != NULL) tail = temp;
    return head;
}

bool Remove(node head, node tail, int value) {
    node p, temp;
    if(head == NULL) return false;
    p = head;
    if(p->data == value) {
        if(head == tail){
            head = NULL;
            tail = NULL;
        } else {
            head = head.next
            return true
        }
    }
    while(p->next != null and p->data != value) {
        p = p->next
    }
    if(p -> next != null) {
        if(p->next == tail) {
            tail = p;
        }
        p->next = p->next->next;
        return true
    }
    return false;
}

File: algorithms/LinkedList.go
package main

import "fmt"

type Node Struct {
    prev *Node
    next *Node
    key interface{}
}

type List struct {
    head *Node
    tail *Node
}

fun (L *List) Insert(key interface{}) {
    list := &Node {
        next: L.head,
        key: key,
    }
    if L.head != nil {
        L.head.prev = list
    }
    L.head = list

    l := L.head
    for l.next != nil {
        l = l.next
    }
    L.tail = l
}

fun Display(list *Node) {
    for list != nil {
        fmt.Printf("%v <-", list.key)
        list = list.prev
    }
    fmt.Println()
}

func (l *List) Reverse() {
    curr := l.head
    var prev *Node
    l.tail = l.head

    for curr != nil {
        next := curr.next
        curr.next = prev
        prev = curr
        curr = next
    }
    l.head = prev
    Display(l.head)
}

func main() {
	link := List{}
	link.Insert(5)
	link.Insert(9)
	link.Insert(13)
	link.Insert(22)
	link.Insert(28)
	link.Insert(36)
	
	fmt.Println("\n==============================\n")
	fmt.Printf("Head: %v\n", link.head.key)
	fmt.Printf("Tail: %v\n", link.tail.key)
	link.Display()
	fmt.Println("\n==============================\n")
	fmt.Printf("head: %v\n", link.head.key)
	fmt.Printf("tail: %v\n", link.tail.key)
	link.Reverse()
	fmt.Println("\n==============================\n")
}